### DISCUSSION TITLE
Welcome to CodeOcean Discussions!

### DISCUSSION THREAD

**Discussion by MrSerth** [openHPI/codeocean#2651] [Announcements]
## 👋 Welcome to our Community Discussions!

We’re excited to use Discussions as a space for all of us—developers, educators, and contributors—to connect, share, and support each other. Here, you can:

* **Ask questions** about any challenges you’re facing.
* **Share your ideas** and project insights.
* **Engage with others**—give feedback, offer advice, and learn from your peers.
* **Welcome new members** and foster a positive, collaborative environment.

## 🔗 Scope of Discussions

This place is dedicated to questions, tips, and ideas about everything related to using, developing, and administrating CodeOcean. For example, you can use this place for anything related to:

- **[Poseidon](https://github.com/openhpi/poseidon)**: Engage about our custom code execution middleware, configuration difficulties or monitoring insights.
- **[Dockerfiles](https://github.com/openhpi/dockerfiles)**: Share best practices or troubleshoot image configurations.
- **[Nomad](https://www.nomadproject.io)**: Ask about job scheduling, workload management, and best-practices for scaling your deployment.
- **[Docker](https://www.docker.com/)**: Discuss anything Docker-related, from setup to optimization.
- **[gVisor](https://gvisor.dev/)**: Understand the underlying security architecture.

For bug reports and feature requests, please continue to open Issues on the respective repositories to keep everything organized.

## 📢 Purpose of Discussions

Use this space to seek peer-to-peer support and exchange knowledge with fellow developers and instructors. Feel free to ask questions, share your experiences, or dive into ongoing conversations. Here are some examples of what to discuss:

* 💻 **Developer-Related Questions**: E.g., "How do I implement a new feature?"
* ⚙️ **Setup Guidance**: E.g., "What’s the best way to set up a production environment for CodeOcean?"
* 🚀 **Performance Issues**: E.g., "Why are code executions running slower than expected?"
* 🧑‍🏫 **Exercise Design**: E.g., "Any tips on structuring exercises to maximize engagement?"
* 📅 **Roadmap Updates**: Stay tuned for updates on features, releases, and community initiatives.
* 🌍 **Community Planning**: Help shape the future of CodeOcean with your ideas and feedback.

We’ll also share updates and roadmap ideas here, so be sure to stay informed by **watching this discussion** (learn how [here](https://docs.github.com/en/account-and-profile/managing-subscriptions-and-notifications-on-github/setting-up-notifications/configuring-notifications#configuring-your-watch-settings-for-an-individual-repository)).

## 📝 Get Involved

This community is for everyone, and we’re building it together 💪. To get started, introduce yourself in the comments below! Let us know what you’re working on, your questions, or any tips you have. Welcome to the CodeOcean community!

**Comment by masterElmar** [comment]
Hello HPI Developer Team and CodeOcean Community,

I'm a student from HTWK Leipzig, and last semester a collegue and I focused on developing Dockerfiles, Compose files, and simplified setup configurations for local use of CodeOcean. Our goal was to make CodeOcean components easily accessible and deployable in containerized environments. We envisioned scaling these setups within a Kubernetes-based MOOC environment.

One of our main objectives was to create a new adapter for Poseidon that could connect to Kubernetes, enabling it to run desired images within pods. However, due to time constraints and the complexity of abstracting between Nomad and Kubernetes, we couldn’t fully implement this functionality.

We're also excited to see new components written in Golang—it’s great to see CodeOcean evolving!
Looking forward to discussing this further, and any advice or insights are much appreciated.

Best regards,
Elmar 🧑‍💻

**Reply by MrSerth** [reply]
Hi Elmar,

it's great to have you onboard and I am thrilled to learn about your use case and the work you've invested. Going forward to a containerized environment is awesome, I like that! 

Recently, we added a Dockerfile for production deployments to the repo and also improved the setup documentation. Still, a Nomad cluster is best deployed on VMs outside of containers to avoid any docker-in-docker setups.

Integrating Kubernetes sounds awesome, too! We never tried, but considered Kubernetes before deciding for Nomad. Mainly, our decision was driven by performance considerations and an evaluation we performed three years ago. Since technology evolves quickly, revisiting this decision is valid and I would be happy to lern more about the experience you made. Our evaluation results are published: 

> **Improving the Scalability and Security of Execution Environments for Auto-Graders in the Context of MOOCs**
> Sebastian Serth, Daniel Köhler, Leonard Marschke, Felix Auringer, Konrad Hanff, Jan-Eric Hellenberg, Tobias Kantusch, Maximilian Paß, and Christoph Meinel
> _In Proceedings of the Fifth Workshop "Automatische Bewertung von Programmieraufgaben" (ABP 2021), Oct 2021_
> DOI: [10.18420/abp2021-1](https://doi.org/10.18420/abp2021-1)
> URL: https://dl.gi.de/handle/20.500.12116/37539

Furthermore, we also looked into using Serverless Functions as an additional code execution environment recently. You'll find more results online, too:

> **On the Feasibility of Serverless Functions in the Context of Auto-Graders**
> Sebastian Serth, Maximilian Paß, and Christoph Meinel
> _In 2023 IEEE 2nd German Education Conference (GECon), Aug 2023_
> DOI: [10.1109/GECon58119.2023.10295106](https://doi.org/10.1109/GECon58119.2023.10295106)
> URL: https://ieeexplore.ieee.org/document/10295106/

I hope some of these resources might be useful for you! Feel free to share additional insights or results 🙂

Best regards,
Sebastian

**Comment by seemantTUI** [comment]
Hello,

I am currently working on my master's thesis at Technical University Ilmenau, where I am trying to automate the request for comments process by using ChatGPT. For example, when a user requests a comment for their exercise, an HTTP request is made to ChatGPT with the user's question, test results, etc. Then, a comment is automatically created on the back-end with a response from ChatGPT. The user can then see this comment on the Request for Comments page, where they will see the response from ChatGPT as a comment.

**Reply by MrSerth** [reply]
Hi @seemantTUI,

nice to have you onboard! I really like your idea of adding ChatGPT to RfCs and think this will be a great addition for learners and teachers alike.

I've extracted your question to a dedicated thread (to have one discussion per topic).

Best,
Sebastian
